modin/core/dataframe/pandas/dataframe/dataframe.py

```diff
@@ -1978,13 +1980,109 @@ def sort_by(
+        # None. This fixes the indexes name beforehand in that case, so that the sort works.
+
+        def sort_function(df):
+            index_renaming = None
+            if any(name in df.columns for name in df.index.names):
+                index_renaming = df.index.names
+                df.index = df.index.set_names([None] * len(df.index.names))
+            df = df.sort_values(by=columns, ascending=ascending, **kwargs)
+            if index_renaming is not None:
+                df.index = df.index.set_names(index_renaming)
+            return df
+
+        axis = Axis(axis)
+        if axis == Axis.ROW_WISE:
+            # If this df is empty, we don't want to try and shuffle or sort.
+            if len(self.axes[0]) == 0 or len(self.axes[1]) == 0:
+                return self.copy()
+            # If this df only has one row partition, we don't want to do a shuffle and sort - we can
+            # just do a full-axis sort.
+            if len(self._partitions) == 1:
+                return self.apply_full_axis(
+                    1,
+                    sort_function,
+                )
+            if self.dtypes[columns[0]] == object:
+                # This means we are not sorting numbers, so we need our quantiles to not try
+                # arithmetic on the values.
+                method = "inverted_cdf"
+            else:
+                method = "linear"
+            from .utils import build_sort_functions
+
+            shuffling_functions = build_sort_functions(
+                self,
+                columns[0],
+                method,
+                ascending if not isinstance(ascending, list) else ascending[0],
+                **kwargs,
+            )
+            major_col_partition_index = self.columns.get_loc(columns[0])
+            cols_seen = 0
+            index = -1
+            for i, length in enumerate(self._column_widths_cache):
+                cols_seen += length
+                if major_col_partition_index <= cols_seen:
+                    index = i
+                    break
+            new_partitions = self._partition_mgr_cls.shuffle_partitions(
+                self._partitions,
+                index,
+                shuffling_functions.sample_function,
+                shuffling_functions.pivot_function,
+                shuffling_functions.split_function,
+                sort_function,
+            )
+            new_axes = self.axes
+            if kwargs.get("ignore_index", False):
+                new_axes[axis.value] = RangeIndex(len(new_axes[axis.value]))
+            else:
+                new_axes[axis.value] = self._compute_axis_labels_and_lengths(
+                    axis.value, new_partitions
+                )[0]
+            # Sometimes when performing a sort for `value_counts` the partitions have a MultiIndex
+            # but the _compute_axis_labels_and_lengths function gives us an Index of tuples instead
+            # of a MultiIndex.
```

Comment:
this sounds like a workaround (at least the comment sounds like that), what is the reason for `value_counts` to do some stuff _sometimes_?

Reply:
I believe this is happening when some of the partitions are empty - if we append a multi index to an empty index, we get a flattened empty index, so if the first partition is empty (which may happen if pivots are poor, or the df is very small) we will get a flattened index. I can add a comment to explain this!
```python
In [17]: from pandas import Index

In [18]: from pandas import MultiIndex

In [20]: a = MultiIndex.from_tuples([(1, 2), (3, 4)])

In [25]: Index([]).append([a])
Out[25]: Index([(1, 2), (3, 4)], dtype='object')
```

Reply:
yeah, some explanatory comment would be awesome